### DISCUSSION TITLE
How to convert mat3(mvi) in GLSL into WGSL?

### DISCUSSION THREAD

**Discussion by jack1232** [gpuweb/gpuweb#1421] [Q&A]
I have the following GLSL code in the fragment shader when doing lighting:

mat4 mvi = transpose(inverse(modelViewMatrix));
vec3 my_normal = mat3(mvi) * normal;

How to convert  this code into WGSL code?

**Comment by kvark** [comment]
Note that there isn't any `inverse` today - https://gpuweb.github.io/gpuweb/wgsl.html#matrix-builtin-functions

**Comment by kainino0x** [comment]
Consider uploading the inverse matrix separately. Computing the inverse in-shader is costly and unnecessarily happens once in every shader invocation.

**Reply by kvark** [reply]
That's not the answer to the question though. It's about taking mat3 part out of mat4. And in that sense, it's a real spec issue, I don't think it should have been moved into a discussion.

**Reply by kainino0x** [reply]
I didn't pick up on that, since the question was generally "how do I convert this" (and as pointed out the inverse() also can't be directly translated). I think that should be filed as an issue separately from this.

**Reply by kainino0x** [reply]
Also sorry this "answer" ended up as an _answer_, I posted it before converting to discussion. Wanted to change it to a comment on the question but that doesn't appear to be a thing.

**Comment by mikialex** [comment]
Is there any update on spec that directly allow converting mat3 from mat4? I'm translating similar code like above and feel so ridiculous...
